File: Algo/Leetcode/230KthSmallestNodeInBST.cpp
// dfs
class Solution {
public:
    int res = 0;
    int kthSmallest(TreeNode* root, int k) {
        dfs(root, k);
        return res;
    }

    void dfs(TreeNode* root, int& count){
        if(!root || count < 0) return; // Do not need to go too deep

        dfs(root->left, count);
        count--;
        if(count == 0){
            res = root->val;
            return;
        }
        dfs(root->right, count);
    }
};

File: Algo/Leetcode/230KthSmallestNodeInBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def __init__(self):
        self.count = 0
        self.res = 0

    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        if root is None:
            return -1

        self.dfs(root, k)
        return self.res

    def dfs(self, root, k):
        if root is None:
            return

        self.dfs(root.left, k)
        self.count += 1
        if self.count == k:
            self.res = root.val
        self.dfs(root.right, k)
